src/requests/create-signature.ts

```diff
@@ -0,0 +1,107 @@
+import * as crypto from 'crypto'
+import type {
+  CanonicalRequest,
+  NormalizedCanonicalRequest,
+  NormalizedHeaders,
+  Secret,
+  Timestamp,
+} from './typings'
+import { CanonicalRequestValidator, SecretValidator, TimestampValidator } from './typings'
+import { getNormalizedEncodedURI, getNormalizedHeaders } from './utils'
+import { ContentfulSigningHeader } from './constants'
+
+const hash = (normalizedCanonicalRequest: NormalizedCanonicalRequest, secret: string) => {
+  const stringifiedHeaders = normalizedCanonicalRequest
+    .headers!.map(([key, value]) => `${key}:${value}`)
+    .join(';')
+
+  const stringifiedRequest = [
+    normalizedCanonicalRequest.method,
+    normalizedCanonicalRequest.path,
+    stringifiedHeaders,
+    normalizedCanonicalRequest.body,
+  ].join('\n')
+
+  const hmac = crypto.createHmac('sha256', secret)
+
+  hmac.update(stringifiedRequest)
+
+  return hmac.digest('hex')
+}
+
+const enrichNormalizedHeadersWithMetadata = (headers: NormalizedHeaders, timestamp: number) => {
+  const result = [...headers]
+  const joinedSignedHeaders = headers.map(([key]) => key).join(',')
+
+  result.push([ContentfulSigningHeader.Timestamp, timestamp.toString()])
+
+  if (joinedSignedHeaders) {
+    result.push([ContentfulSigningHeader.SignedHeaders, joinedSignedHeaders])
+  }
+
+  return result
+}
+
+/**
+ * Given a secret, a canonical request and a timestamp, generates a signature.
+ * It can be used to verify canonical requests to assess authenticity of the
+ * sender and integrity of the payload.
+ *
+ * ~~~
+ * const {createSignature} = require('contentful-node-apps-toolkit')
+ * const {pick} = require('lodash')
+ * const {server} = require('./imaginary-server')
+ *
+ * const SECRET = process.env.SECRET
+ *
+ * server.post('/api/my-resources', (req, res) => {
+ *   const incomingSignature = req.headers['x-contentful-signature']
+ *   const incomingTimestamp = req.headers['x-contentful-timestamp']
+ *   const now = Date.now()
+ *
+ *   if (!incomingSignature) {
+ *     res.send(400, 'Missing signature')
+ *   }
+ *
+ *   if (now - incomingTimestamp > 1000) {
+ *     res.send(408, 'Request too old')
+ *   }
+ *
+ *   const computedSignature = createSignature(
+ *     SECRET,
+ *     {
+ *       method: req.method,
+ *       path: req.url,
+ *       headers: pick(req.headers, ['Authorization']),
```

Comment:
I would show how you pick them using x-contentful-signed-headers

Reply:
That part is in `is-verified-request`, I thought about giving a different example here. Maybe better be consistent here though, so I will change it